overlays/haskell.nix

```diff
@@ -648,57 +653,97 @@ final: prev: {
         # separately from the hsPkgs.  The advantage is that the you can get the
         # plan-nix without building the project.
-        cabalProject' =
-          projectModule: haskellLib.evalProjectModule ../modules/cabal-project.nix projectModule (
-            { src, compiler-nix-name, compilerSelection, evalPackages, ... }@args:
-            let
-              callProjectResults = callCabalProjectToNix args;
-              plan-pkgs = importAndFilterProject {
-                inherit (callProjectResults) projectNix sourceRepos src;
-              };
-              buildProject = if final.stdenv.hostPlatform != final.stdenv.buildPlatform
-                then final.buildPackages.haskell-nix.cabalProject' projectModule
-                else project;
-              pkg-set = if plan-pkgs ? configurationError
-                then {
-                  inherit (plan-pkgs) configurationError;
-                  config = {
-                    compiler.nix-name = compiler-nix-name;
-                    hsPkgs = {};
-                    inherit evalPackages;
-                  };
-                }
-                else mkCabalProjectPkgSet
-                { inherit compiler-nix-name compilerSelection plan-pkgs;
-                  pkg-def-extras = args.pkg-def-extras or [];
-                  modules = [ { _module.args.buildModules = final.lib.mkForce buildProject.pkg-set; } ]
-                    ++ (args.modules or [])
-                    ++ [ {
-                      ghc.package =
-                        if args.ghcOverride != null
-                          then args.ghcOverride
-                        else if args.ghc != null
-                          then args.ghc
-                        else
-                          final.lib.mkDefault (args.compilerSelection final.buildPackages).${compiler-nix-name};
-                      compiler.nix-name = final.lib.mkForce args.compiler-nix-name;
-                      evalPackages = final.lib.mkDefault evalPackages;
-                    } ];
-                  extra-hackages = args.extra-hackages or [] ++ callProjectResults.extra-hackages;
+        cabalProject' = projectModule:
+          (final.lib.evalModules {
+            modules =
+            (if builtins.isList projectModule then projectModule else [ projectModule ]) ++
+            [
+              (import ../modules/project-common.nix)
+              (import ../modules/cabal-project.nix)
+              # Pass the pkgs and the buildProject to the modules
+              ({ config, lib, ... }: {
+                _module.args = {
+                  pkgs = final;
+                  # to make it easy to depends on build packages in, eg., shell definition:
+                  inherit (config.result-of-f) buildProject;
                 };
+                inherit (config.result-of-f) hsPkgs;
+              })
+              # FIXME: should this be part of cabal-project.nix?
```

Comment:
That would be nice

Reply:
We did move most of the code for `hackage-project` into a module file and I think it is nicer like that.